1-Draft/RFC00XX-PredIctive-IntelliSense.md

```diff
@@ -0,0 +1,359 @@
+receiving an updated prediction.
+
+![arg change](./media/pi-fb.gif)
+
+## Prediction Extension Model
+
+Additional providers enhance historic predictions by providing domain specific commands and task
+completions. Predictive IntelliSense includes an extension model to support the registration of
+additional providers.
+
+## Goals/Non-Goals
+
+ Goals:
+
+* Provide extension model to support additional providers
+* Support the discovery of providers using Get-Subsystem
+* Render predictions from providers within performance guidelines 20ms
+* Provide Module owners method of registration
+
+Non-goals:
+
+* Register/Unregister-Subsystem cmdlets are not planned because it's targeting binary subsystem
+  implementations, which should deal with registration/unregistration via APIs.
+* Server-side provider privacy is the responsibility of the provider module
+
+## Extension Model Specification
+
+### Predictive Suggestion Contract to Support Addition Providers
+
+Below is an abbreviated code listing of the API contract to support suggestions from additional
+providers.
+
+```csharp
+/// The class represents a predictive suggestion generated by a predictor.
+public sealed class PredictiveSuggestion
+{
+/// Gets the suggestion.
+public string SuggestionText { get; }
+
+/// Gets the tooltip of the suggestion.
+public string? ToolTip { get; }
+
+/// Initializes a new instance of the <see cref="PredictiveSuggestion"/> class.
+/// <param name="suggestion">The predictive suggestion text.</param>
+public PredictiveSuggestion(string suggestion)
+: this(suggestion, toolTip: null)
+```
+
+### Discovery of Provider Metadata and Registration
+
+Providers are registered automatically when the user installs a provider module. Removal of
+registration occurs when the module is removed. Users will need to discover installed providers
+along with additional metadata, such as version information to assist users in upgrading and
+diagnosing problems.
+
+To list the currently installed providers and their respective metadata:
+
+```powershell
+Get-SubSystem
+```
+
+### Expected Provider Performance
+
+Additional providers often will communicate with an internet connect service to provide suggestions.
+To minimize the appearance of slow responses, we estimate that the maximum roundtrip time of 20ms is
+permissible to not negatively impact users.
+
+If the roundtrip time is exceeded, the predictor model will provide historical suggestions, or
+suggestions from another provider that completed in the allowed time.
+
+* Default required provider performance = 20ms
+
+The current roundtrip time in PSReadLine using the Tab Completion feature is approximately **2-3ms**. We
+believe that this current default roundtrip time for additional providers may still exceed user
+expectations. With additional testing and feedback, we may need to reduce the roundtrip time to less
+than **20ms**.
+
+### Registration for Module Owners
```

Comment:
Need to update to include the data sent back from PSReadLine to the provider to indicate acceptance of a prediction so that the provider can refine it's results next time

Reply:
Discussed with @daxian-dbw and removing this section from this RFC. Will be included with implementation and design details in Dongbo's RFC.